### PR TITLE
feat(audit-trail): add support for audit trail to terraform provider

### PR DESCRIPTION
**Pull request by mwasilew2**
## Description of the change

closes: https://app.clickup.com/t/8693fq8pf

blocked on:
- ~expose id~ won't be needed:
    - ~https://github.com/spacelift-io/backend/pull/6147~
    - ~https://github.com/spacelift-io/backend/pull/6165~
- https://github.com/spacelift-io/backend/pull/6154

## Type of change

- [ ] Bug fix (non-breaking change that fixes an issue)
- [ ] New feature (non-breaking change that adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation (non-breaking change that adds documentation)

## Related issues

> Fix [#1]()

### Development

- [ ] Lint rules pass locally
- [ ] The code changed/added as part of this pull request has been covered with tests
- [ ] All tests related to the changed code pass in development
- [ ] Examples for new resources and data sources have been added
- [ ] Default values have been documented in the description (e.g., "Dummy: (Boolean) Blah blah. Defaults to `false`.)
- [ ] If the action fails that checks the documentation: Run `go generate` to make sure the docs are up to date

### Code review

- [ ] This pull request has a descriptive title and information useful to a reviewer. There may be a screenshot or screencast attached
- [ ] Pull Request is no longer marked as "draft"
- [ ] Reviewers have been assigned
- [ ] Changes have been reviewed by at least one other engineer

### CODE REVIEW

**Comment by Clickup-user** [comment]
Task linked: [CU-8693fq8pf add support for `auditTrailWebhook` to terraform provider](https://app.clickup.com/t/8693fq8pf)

**Comment by mwasilew2** [comment]
@peterdeme @michalg9 I saw you working on terraform provider recently, you would mind taking a look?

File: spacelift/resource_audit_trail_webhook.go [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+			Endpoint:    toString(data.Get("endpoint")),
+			IncludeRuns: toBool(data.Get("include_runs")),
+			Secret:      toString(data.Get("secret")),
+		},
+	}
+	if err := i.(*internal.Client).Mutate(ctx, "AuditTrailWebhookCreate", &mutation, variables); err != nil {
+		return diag.Errorf("could not create audit trail webhook: %v", internal.FromSpaceliftError(err))
+	}
+
+	data.SetId(time.Now().String())
```
[the remark this answers is not in the document]
Comment (mwasilew2) [reply spacelift/resource_audit_trail_webhook.go:65]: @peterdeme 

nope, not at all, I mean it works, but it's a hack, see: https://spacelift-io.slack.com/archives/CRTS17WLF/p1705055720419239

[the remark this answers is not in the document]
Comment (mwasilew2) [reply spacelift/resource_audit_trail_webhook.go:65]: I also drafted PRs to actually execute the "ulid based" solution: https://github.com/spacelift-io/backend/pull/6165

[the remark this answers is not in the document]
Comment (peterdeme) [reply spacelift/resource_audit_trail_webhook.go:65]: I see. I don't have a strong opinion on this, so feel free to do whatever :D Personally, I would either generate a "smart" id (maybe from endpoint URL?), or go through with the ULID pull request.

Waiting on other people's opinion.

**Comment by mwasilew2** [comment]
@marcinwyszynski @adamconnelly I saw you've been active in the repo, could you take a look? Thanks in advance

**Review by adamconnelly — COMMENTED** [review]
Looks good overall. Left some comments, but nothing major.

File: spacelift/resource_audit_trail_webhook_test.go [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+			},
+			{
+				ResourceName:      resourceName,
+				ImportState:       true,
+				ImportStateVerify: true,
+			},
+		})
+	})
+
+	t.Run("cannot change endpoint", func(t *testing.T) {
```
[the remark this answers is not in the document]
Comment (mwasilew2) [reply spacelift/resource_audit_trail_webhook_test.go:45]: > testing that the endpoint can't be changed (which should totally be possible)?

It's not possible to edit the endpoint if the audit trail webhook is enabled.

File: spacelift/resource_audit_trail_webhook_test.go [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+var auditTrailWebhookSimple = `
+resource "spacelift_audit_trail_webhook" "test" {
+	enabled = true
+	endpoint = "%s"
+	include_runs = true
+	secret = "secret"
+}
+`
+
+func Test_resourceAuditTrailWebhook(t *testing.T) {
```
[the remark this answers is not in the document]
Comment (mwasilew2) [reply spacelift/resource_audit_trail_webhook_test.go:22]: Deletion should be a part of each test (the test will fail if the resource created for the test was not successfully removed). Do you reckon we need to test it explicitly? Do you know how to test it explicitly? I don't see a test like that in the repo and I'm not sure how to do it.

[the remark this answers is not in the document]
Comment (adamconnelly) [reply spacelift/resource_audit_trail_webhook_test.go:22]: That's fine then - don't worry about it if the test fails anyway if deletion fails. I just hadn't realised that.

As for how you'd do it - I imagine you'd just add a test step that applied a configuration where the resource wasn't included.

[the remark this answers is not in the document]
Comment (mwasilew2) [reply spacelift/resource_audit_trail_webhook_test.go:22]: > As for how you'd do it - I imagine you'd just add a test step that applied a configuration where the resource wasn't included.

yeah, just wasn't sure how to test for a missing resource after applying an empty config

File: spacelift/resource_audit_trail_webhook_test.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (mwasilew2) [reply spacelift/resource_audit_trail_webhook_test.go:45]: you're right: https://github.com/spacelift-io/backend/blob/c7dbd46d8f09fbee7b95da3c2ac44e46c85d3db2/server/resolvers/audit_trail_webhook_mutations.go#L74

I think the reason I was able to set the endpoint to a non-existent value was that I disabled audit trail first and the check is only performed if it's enabled.

thanks!

I think the test still brings some value so I'll leave it and update the name.

**Pull request merged**